### ISSUE TITLE
master, layer-opacity and mod_tile

### ISSUE THREAD

**Issue by Andrey-VI** [mapnik/mapnik-support#21]
1. With current Mapnik master I get error merging https://github.com/mapnik/mapnik/tree/layer-opacity branch in it:

``` sh
andrey@cart-dev:~/sources$ git clone https://github.com/mapnik/mapnik.git -b layer-opacity layer-opacity
andrey@cart-dev:~/sources$ cd layer-opacity
andrey@cart-dev:~/sources/layer-opacity$ git merge remotes/origin/master
Removing tests/cpp_tests/svg_renderer_tests/path_element_test.cpp
Removing tests/cpp_tests/svg_renderer_tests/build.py
Auto-merging src/load_map.cpp
Auto-merging src/layer.cpp
CONFLICT (content): Merge conflict in src/layer.cpp
Auto-merging include/mapnik/view_transform.hpp
Auto-merging include/mapnik/layer.hpp
Auto-merging include/mapnik/agg_renderer.hpp
Automatic merge failed; fix conflicts and then commit the result.
```

So I hope that https://github.com/mapnik/mapnik/issues/1818 will land in master soon.
cc @artemp 
1. mod_tile unable to load XMLs when using Mapnik 3.x master. It just uses all available system memory (including swap) and I have no other choice but to kill the process.:
   ![_mod_tile](https://cloud.githubusercontent.com/assets/1121499/4282652/dfb84704-3d5d-11e4-89cd-b79a9a24b81b.PNG)

**Comment by springmeyer** [comment]
I just fixed the merge of master into layer-opacity. Can you try again?

Also, are you using my branch of mod_tile? https://github.com/springmeyer/mod_tile

**Comment by Andrey-VI** [comment]
Yes, I used your patches to mod_tile before and this branch also. They are almost identical.

> I just fixed the merge of master into layer-opacity. Can you try again?

Thanks. Finally, this Mapnik version works with your mod_tile branch.
But I get errors like:
`Unknown element 'ExpressionFormat' in style 'country' in ExpressionFormat at line 3054 of`
No good. This means that I can't use same XMLs with both versions of Mapnik (2.3.x and 3.x).

Also I noticed that the Mapnik 3.x requires a lot of memory (around 34 GiB) during compiling. Is this OK?

**Comment by springmeyer** [comment]
> Thanks. Finally, this Mapnik version works with your mod_tile branch.

Great.

> But I get errors like:
> Unknown element 'ExpressionFormat' in style 'country' in ExpressionFormat at line 3054 of
> No good. This means that I can't use same XMLs with both versions of Mapnik (2.3.x and 3.x).

Can you create a ticket for this and I will fix asap (adding a back compatibility fix)?

> Also I noticed that the Mapnik 3.x requires a lot of memory (around 34 GiB) during compiling. Is this OK?

Not that is not okay. Can you tell which cpp file is requiring that much memory? Also, what compiler version are you using? Are you compiling in parallel? (Note: the `make` wrapper around scons does by default)

**Comment by Andrey-VI** [comment]
> Can you tell which cpp file is requiring that much memory?

Good question… Actually, I don't remember anything special. I think all threads used ~5 % of memory.

> Also, what compiler version are you using?

Debian default.

``` sh
andrey@cart-dev:~/sources$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 4.9.1-4' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --with-arch-32=i586 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.1 (Debian 4.9.1-4)
```

> Are you compiling in parallel? (Note: the  make  wrapper around scons does by default)

Yes, 24 parallel jobs. But I have no such problem with 2.3.x branch.

**Comment by springmeyer** [comment]
Okay, 24 parellel jobs is a lot with Mapnik 3.x (which does require more memory that 2.3.x). So it may be that no single cpp file is requiring that much memory but rather than many cpp files that are concurrently being built all require 1-2 GB each, which could quickly add up. If that is what is happening then I'd say 34 GB might be expected behavior.

One thing to consider is using clang++ instead of g++. Clang++ uses less memory than g++ in many cases and produces binaries that should run just as fast and will be compatible with other code compiled with g++. I use clang++ exclusively these days on both osx and linux. On linux I've been using clang 3.4 and 3.5 via the llvm deb repos: http://llvm.org/apt/

**Comment by Andrey-VI** [comment]
> Okay, 24 parellel jobs is a lot with Mapnik 3.x (which does require more memory that 2.3.x).

Yes. I tried 16 threads (`./configure JOBS=16`), but it still uses 24. Is this bug?

> One thing to consider is using clang++ instead of g++.

OK, I'll take a look at clang.

**Comment by Andrey-VI** [comment]
I have clang 3.4 installed. How to force it to build Mapnik?

**Comment by springmeyer** [comment]
> Yes. I tried 16 threads (./configure JOBS=16), but it still uses 24. Is this bug?

Try: `JOBS=16 make`

> I have clang 3.4 installed. How to force it to build Mapnik?

```
./configure CXX=/usr/bin/clang++ CC=/usr/bin/clang
```

Or point it at `/usr/bin/clang++-3.4` if that is what you have...

**Comment by Andrey-VI** [comment]
OK, thanks!

**Comment by Andrey-VI** [comment]
With `./configure CXX=clang++ CC=clang` I have no memory problems when compiling even with 24 jobs. The process uses ~12 GiB, but it's OK for me. 
process_markers_symbolizer.cpp requires ~5 GiB of RAM in 3 threads.

**Comment by Andrey-VI** [comment]
Hm… I just want to test new `horizontal-alignment="adjust"`. I recompiled Mapnik and mod_tile, and faced again with memory leak problem when starting renderd…
Don't know what to do.

**Comment by springmeyer** [comment]
> Yes. I tried 16 threads (./configure JOBS=16), but it still uses 24. Is this bug?

Fixed in https://github.com/mapnik/mapnik/issues/2430

> Hm… I just want to test new horizontal-alignment="adjust". I recompiled Mapnik and mod_tile, and faced again with memory leak problem when starting renderd…
> Don't know what to do.

It sounds like perhaps the re-compile is not fully working and there are some stale binaries somehow? Maybe you need to do a full clean of mod_tile?

**Comment by Andrey-VI** [comment]
> Fixed in mapnik/mapnik#2430

Thanks.

> It sounds like perhaps the re-compile is not fully working and there are some stale binaries somehow? Maybe you need to do a full clean of mod_tile?

I compiled mod_tile from clean sources. No luck.
Switching back to 2.3.x always worked OK.

Is there any way to track down the cause of renderd memory leaks?

**Comment by springmeyer** [comment]
> Is there any way to track down the cause of renderd memory leaks?

Yes, install valgrind, and then run `valgrind --leak-check=full ./renderd` and post the results so we can dissect. I would normally offer to try to replicate but renderd is such a pain to run on OS X that I'm not going to try to set it up locally.

**Comment by Andrey-VI** [comment]
Preparing to install valgrind I updated my Debian testing. After that I recompiled Mapnik and mod_tile again. And renderd starts normally now.
So I don't know what the problem was. Last time I updated my system two or three months ago.

**Comment by springmeyer** [comment]
Okay, sounds good, so let's close this then and just re-open a new ticket if you hit any further issues.
